File: go/tests/error_paths_test.go
package tests

// TC3 — Error/validation paths with TYPED errors. Mirrors the C# TestErrorPaths
// suite. Go surfaces errors as values, so we classify via errors.As into the
// generated *airtable.APIError (which carries StatusCode + Type, where Type is
// the Airtable error code such as NOT_FOUND) rather than asserting a bare
// err != nil. This proves the client classifies HTTP failures into specific
// typed errors instead of one opaque error. Parity target.

import (
	"context"
	"errors"
	"testing"
	"time"

	airtable "myairtabletests/output"
)

// newAirtableWithBadKey builds a client pointed at the REAL base but with a
// bogus PAT, so the auth (401) path is exercised rather than the base (404).
// Mirrors C# TestSetup.MakeAirtableWithBadKey.
func newAirtableWithBadKey(t *testing.T) *airtable.Airtable {
	t.Helper()
	requireCreds(t)
	return airtable.NewWithBase(bogusAPIKey, baseID(), 0)
}

// asAPIError asserts err is a typed *airtable.APIError and returns it. It fails
// the test (with the actual concrete type) when the error is not classified —
// the whole point of TC3 is that failures are typed, not opaque.
func asAPIError(t *testing.T, err error) *airtable.APIError {
	t.Helper()
	if err == nil {
		t.Fatal("expected an error, got nil")
	}
	var apiErr *airtable.APIError
	if !errors.As(err, &apiErr) {
		t.Fatalf("expected *airtable.APIError, got %T: %v", err, err)
	}
	return apiErr
}

func errorPathPrimaryFields(suite string) *airtable.Fields {
	fields := airtable.NewFields(nil, airtable.PrimaryNameToID)
	_ = fields.Set(airtable.PrimaryPrimaryKeyFieldID, suite)
	return fields
}

func TestErrorPaths(t *testing.T) {
	at := newAirtable(t)
	ctx := context.Background()

	t.Run("GetNonexistentRecord", func(t *testing.T) {
		_, err := at.PrimaryDict.GetOne(ctx, "recDOESNOTEXIST0001")
		api := asAPIError(t, err)
		t.Logf("404 -> *APIError{StatusCode:%d, Type:%q}: %v", api.StatusCode, api.Type, err)
		if api.StatusCode != 404 {
			t.Fatalf("expected 404, got %d", api.StatusCode)
		}
		// The 404 sentinel must also match for this typed error.
		if !errors.Is(err, airtable.ErrNotFound) {
			t.Fatalf("expected errors.Is(err, ErrNotFound) for a 404")
		}
		if api.Type != "NOT_FOUND" {
			t.Fatalf("expected code NOT_FOUND, got %q", api.Type)
		}
	})

	t.Run("CreateWithInvalidSelectOption", func(t *testing.T) {
		fields := errorPathPrimaryFields(primaryKey("Error", "BadSelect"))
		_ = fields.Set(airtable.PrimarySingleSelectFieldID, "NotARealOption_zzz")
		created, err := at.PrimaryDict.CreateOne(ctx, fields)
		if err == nil {
			at.PrimaryDict.DeleteOne(ctx, created.ID) //nolint:errcheck // cleanup unexpected success
			t.Fatal("expected an error creating with an invalid select option")
		}
		api := asAPIError(t, err)
		t.Logf("bad select -> *APIError{StatusCode:%d, Type:%q}: %v", api.StatusCode, api.Type, err)
		if api.Type != "INVALID_MULTIPLE_CHOICE_OPTIONS" {
			t.Fatalf("expected code INVALID_MULTIPLE_CHOICE_OPTIONS, got %q", api.Type)
		}
	})

	t.Run("CreateWithWrongValueType", func(t *testing.T) {
		fields := errorPathPrimaryFields(primaryKey("Error", "BadType"))
		_ = fields.Set(airtable.PrimaryNumberIntFieldID, "not a number")
		created, err := at.PrimaryDict.CreateOne(ctx, fields)
		if err == nil {
			at.PrimaryDict.DeleteOne(ctx, created.ID) //nolint:errcheck // cleanup unexpected success
			t.Fatal("expected an error creating with a wrong-typed value")
		}
		api := asAPIError(t, err)
		t.Logf("wrong type -> *APIError{StatusCode:%d, Type:%q}: %v", api.StatusCode, api.Type, err)
		if api.Type != "INVALID_VALUE_FOR_COLUMN" {
			t.Fatalf("expected code INVALID_VALUE_FOR_COLUMN, got %q", api.Type)
		}
	})

	t.Run("CreateWithUnknownField", func(t *testing.T) {
		fields := errorPathPrimaryFields(primaryKey("Error", "BadField"))
		_ = fields.Set("fldDOESNOTEXIST00000", "x")
		created, err := at.PrimaryDict.CreateOne(ctx, fields)
		if err == nil {
			at.PrimaryDict.DeleteOne(ctx, created.ID) //nolint:errcheck // cleanup unexpected success
			t.Fatal("expected an error creating with an unknown field id")
		}
		api := asAPIError(t, err)
		t.Logf("unknown field -> *APIError{StatusCode:%d, Type:%q}: %v", api.StatusCode, api.Type, err)
		if api.Type != "UNKNOWN_FIELD_NAME" {
			t.Fatalf("expected code UNKNOWN_FIELD_NAME, got %q", api.Type)
		}
	})

	t.Run("BadApiKeyIsAuthError", func(t *testing.T) {
		bad := newAirtableWithBadKey(t)
		_, err := bad.PrimaryDict.GetMany(ctx, (&airtable.Query{}).WithMaxRecords(1))
		api := asAPIError(t, err)
		t.Logf("bad key -> *APIError{StatusCode:%d, Type:%q}: %v", api.StatusCode, api.Type, err)
		if api.StatusCode != 401 {
			t.Fatalf("expected 401, got %d", api.StatusCode)
		}
		if api.Type != "AUTHENTICATION_REQUIRED" {
			t.Fatalf("expected code AUTHENTICATION_REQUIRED, got %q", api.Type)
		}
	})

	t.Run("RateLimitedErrorIsDistinctType", func(t *testing.T) {
		// Go exposes a distinct *airtable.RateLimitError type (429, retries
		// exhausted) carrying its RetryAfter. The live 429 path is covered by
		// TC8; here we assert the type is constructible, distinct from APIError,
		// and carries its Retry-After value.
		err := error(&airtable.RateLimitError{RetryAfter: 30 * time.Second})
		var rl *airtable.RateLimitError
		if !errors.As(err, &rl) {
			t.Fatalf("expected *airtable.RateLimitError, got %T", err)
		}
		if rl.RetryAfter != 30*time.Second {
			t.Fatalf("expected RetryAfter 30s, got %s", rl.RetryAfter)
		}
		// Distinct from APIError — must NOT classify as one.
		var api *airtable.APIError
		if errors.As(err, &api) {
			t.Fatal("RateLimitError must be distinct from APIError")
		}
	})
}

File: typescript/tests/error-paths.test.ts
import { describe, it, expect } from "vitest";
import { Record as ATRecord } from "airtable";
import { Airtable, PrimaryFieldSet, ApiError, HttpError, AirtableError } from "../output";

/**
 * TC3 — Error / validation paths (TYPED).
 *
 * Mirrors csharp/tests/TestErrorPaths.cs now that the generated TypeScript client exposes a typed
 * error hierarchy (myairtable-9yrt): `AirtableError` base with `ApiError` (structured `.code` /
 * `.statusCode` / `.apiMessage`), `HttpError`, `RateLimitedError`, etc. The wrapper classifies HTTP
 * failures instead of flattening them to `new Error(String(error))`.
 *
 * These cases are LIVE — they hit the base and require credentials, so they run in CI, not locally.
 *
 * Two documented divergences from the C# parity target remain, by design:
 *   - Scenario 1 uses a well-formed-but-nonexistent id ("rec" + 14 chars) so it passes the client-side
 *     record-id regex; get-by-id is a filterByFormula query, so a missing record yields zero rows and
 *     the client synthesizes ApiError("NOT_FOUND", 404). (A malformed id is a separate ZodError case.)
 *   - Scenario 4 (unknown field id): the typed writable-field allowlist drops the unknown field
 *     CLIENT-SIDE, so the create SUCCEEDS. This is an intentional client-side-validation divergence,
 *     NOT a regression — the allowlist is deliberately not weakened.
 */

const airtable = new Airtable();

function newPrimaryRecord(): ATRecord<PrimaryFieldSet> {
	return new ATRecord<PrimaryFieldSet>(airtable.primary._table, "", {});
}

describe("TC3 — Error / validation paths", () => {
	// ── Scenario 1: get() a well-formed but nonexistent record id → typed NOT_FOUND ─────────────
	it("get(nonexistent id) throws ApiError NOT_FOUND", async () => {
		const err = await airtable.primary.get("rec00000000000000").then(
			() => {
				throw new Error("expected get() to reject");
			},
			(e) => e,
		);
		expect(err).toBeInstanceOf(ApiError);
		expect(err).toBeInstanceOf(AirtableError);
		expect((err as ApiError).code).toBe("NOT_FOUND");
	}, 30_000);

	// ── Scenario 1b: malformed record id → client-side ZodError (separate case) ──────────────────
	it("get(malformed id) throws a client-side validation error (ZodError, not an ApiError)", async () => {
		const err = await airtable.primary.get("recDOESNOTEXIST0001").then(
			() => {
				throw new Error("expected get() to reject");
			},
			(e) => e,
		);
		expect(err).toBeInstanceOf(Error);
		expect((err as Error).constructor.name).toBe("ZodError");
	}, 30_000);

	// ── Scenario 2: create with an invalid single-select option → INVALID_MULTIPLE_CHOICE_OPTIONS ─
	it("create(invalid select option) throws ApiError INVALID_MULTIPLE_CHOICE_OPTIONS (422)", async () => {
		const r = newPrimaryRecord();
		r.set("Primary Key", `TS Error BadSelect ${Date.now()}`);
		r.set("Single Select", "NotARealOption_zzz" as any);

		const err = await airtable.primary.create(r).then(
			(created) => {
				return airtable.primary.delete((created as any).id).then(() => {
					throw new Error("expected create() to reject");
				});
			},
			(e) => e,
		);

		expect(err).toBeInstanceOf(ApiError);
		expect((err as ApiError).code).toBe("INVALID_MULTIPLE_CHOICE_OPTIONS");
		expect((err as ApiError).statusCode).toBe(422);
	}, 30_000);

	// ── Scenario 3: create with a wrong-typed value → INVALID_VALUE_FOR_COLUMN ───────────────────
	it("create(wrong-typed value) throws ApiError INVALID_VALUE_FOR_COLUMN (422)", async () => {
		const r = newPrimaryRecord();
		r.set("Primary Key", `TS Error BadType ${Date.now()}`);
		r.set("Number (int)", "not a number" as any);

		const err = await airtable.primary.create(r).then(
			(created) => {
				return airtable.primary.delete((created as any).id).then(() => {
					throw new Error("expected create() to reject");
				});
			},
			(e) => e,
		);

		expect(err).toBeInstanceOf(ApiError);
		expect((err as ApiError).code).toBe("INVALID_VALUE_FOR_COLUMN");
		expect((err as ApiError).statusCode).toBe(422);
	}, 30_000);

	// ── Scenario 4: create with an unknown field id → SUCCEEDS (documented divergence) ───────────
	//
	// The typed writable-field allowlist (toWritableRecord keeps only known writable field ids) drops
	// the unknown field id CLIENT-SIDE, so the server never sees it and the create succeeds. This is an
	// intentional client-side-validation divergence from C#'s UNKNOWN_FIELD_NAME (which passes the raw
	// dict straight through), NOT a regression. The allowlist is deliberately not weakened.
	it("create(unknown field id) SUCCEEDS — unknown field dropped client-side (intentional divergence)", async () => {
		const r = new ATRecord<any>(airtable.primary._table, "", {});
		r.set("Primary Key", `TS Error BadField ${Date.now()}`);
		r.set("fldDOESNOTEXIST00000", "x");

		const created = await airtable.primary.create(r);
		const id = (created as any).id as string;
		expect(id).toBeTruthy();
		await airtable.primary.delete(id);
	}, 30_000);

	// ── Scenario 5: real base id + bogus API key → typed 401 ─────────────────────────────────────
	//
	// `new Airtable({ apiKey })` reads the real baseId from env (AIRTABLE_BASE_ID), so this targets auth
	// (401) against the real base. Accept either a structured ApiError(code) or a raw HttpError(401).
	it("bad API key throws a typed 401 (ApiError/HttpError)", async () => {
		const bad = new Airtable({ apiKey: "patBOGUS00000.deadbeefdeadbeefdeadbeefdeadbeef" });

		const err = await bad.primary.get({ maxRecords: 1 }).then(
			() => {
				throw new Error("expected get() to reject");
			},
			(e) => e,
		);

		expect(err).toBeInstanceOf(AirtableError);
		expect(err instanceof ApiError || err instanceof HttpError).toBe(true);
		expect((err as ApiError | HttpError).statusCode).toBe(401);
		if (err instanceof ApiError && err.code) {
			expect(err.code).toBe("AUTHENTICATION_REQUIRED");
		}
	}, 30_000);

	// ── Scenario 6: distinct rate-limited error class ────────────────────────────────────────────
	//
	// A typed `RateLimitedError` (carrying `.retryAfter`) now exists in the generated output. The live
	// 429 path is exercised by TC8 / the retry-5xx suite; no need to force a live 429 here.
});
